cmd/util/ledger/util/registers_util.go

```diff
@@ -0,0 +1,106 @@
+package util
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/rs/zerolog/log"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/onflow/flow-go/cmd/util/ledger/util/registers"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+func NewByAccountRegistersFromPayloadAccountGrouping(
+	payloadAccountGrouping *PayloadAccountGrouping,
+	nWorker int,
+) (
+	*registers.ByAccount,
+	error,
+) {
+	g, ctx := errgroup.WithContext(context.Background())
+
+	jobs := make(chan *PayloadAccountGroup, nWorker)
+	results := make(chan *registers.AccountRegisters, nWorker)
+
+	g.Go(func() error {
+		defer close(jobs)
+		for {
+			payloadAccountGroup, err := payloadAccountGrouping.Next()
+			if err != nil {
+				return fmt.Errorf("failed to group payloads by account: %w", err)
+			}
+
+			if payloadAccountGroup == nil {
+				return nil
+			}
+
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case jobs <- payloadAccountGroup:
+			}
+		}
+	})
+
+	workersLeft := int64(nWorker)
+	for i := 0; i < nWorker; i++ {
+		g.Go(func() error {
+			defer func() {
+				if atomic.AddInt64(&workersLeft, -1) == 0 {
+					close(results)
+				}
+			}()
+
+			for payloadAccountGroup := range jobs {
+
+				// Convert address to owner
+				payloadGroupOwner := flow.AddressToRegisterOwner(payloadAccountGroup.Address)
+
+				accountRegisters, err := registers.NewAccountRegistersFromPayloads(
+					payloadGroupOwner,
+					payloadAccountGroup.Payloads,
+				)
+				if err != nil {
+					return fmt.Errorf("failed to create account registers from payloads: %w", err)
+				}
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case results <- accountRegisters:
+				}
+			}
+
+			return nil
+		})
+	}
+
+	registersByAccount := registers.NewByAccount()
+	g.Go(func() error {
+		for accountRegisters := range results {
+			oldAccountRegisters := registersByAccount.SetAccountRegisters(accountRegisters)
+			if oldAccountRegisters != nil {
+				// Account grouping should never create multiple groups for an account.
+				// In case it does anyway, merge the groups together,
+				// by merging the existing registers into the new ones.
+
+				log.Warn().Msgf(
```

Comment:
It looks like we believe running in this case is an exception, then I would suggest to crash. Instead of merging, I would suggest to do a comparison. 

So we could first print a warning message with a summary, then compare and print detailed comparison result, and then crash. 

I think the detailed crash report would be more useful for understanding and resolving the issue.

Reply:
I remember this specific code block was added by someone else to handle an edge case during migration so it may be risky to modify it this close to code freeze.  

Given this, I added a TODO comment to look into whether the edge case still exists, etc.